### PR TITLE
Modify Kyber KEM EVP implementation

### PR DESCRIPTION
**Pull request by dkostic**
- Remove OpenSSL specific functions.
- Remove OpenSSL style provider implementation.
- Add KEM encapsulate and decapsulate to EVP_PKEY.

### Issues:
Resolves #ISSUE-NUMBER1
Addresses #ISSUE-NUMBER2

### Description of changes: 
Describe AWS-LC’s current behavior and how your code changes that behavior. If there are no issues this pr is resolving,
explain why this change is necessary.

### Call-outs:
Point out areas that need special attention or support during the review process. Discuss architecture or design changes.

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: crypto/evp_extra/p_ed25519.c [hunk 1]
```diff
@@ -104,4 +104,6 @@ const EVP_PKEY_METHOD ed25519_pkey_meth = {
     NULL /* derive */,
     NULL /* paramgen */,
     NULL /* ctrl */,
+    NULL /* encapuslate */,
```
Comment (torben-hansen) [inline crypto/evp_extra/p_ed25519.c:107]: There are a few more places where we define `EVP_PKEY_METHOD`s, where we want the explicit `NULL` e.g. `/crypto/fipsmodule/evp/p_{ec,rsa}.c`

File: crypto/evp_extra/p_ed25519.c [hunk 2]
```diff
@@ -104,4 +104,6 @@ const EVP_PKEY_METHOD ed25519_pkey_meth = {
...
+    NULL /* decapuslate */,
```
Comment (torben-hansen) [inline crypto/evp_extra/p_ed25519.c:108]:
```suggestion
    NULL /* decapsulate */,
```

File: crypto/evp_extra/p_ed25519.c (hunk 1 above)
Comment (torben-hansen) [inline crypto/evp_extra/p_ed25519.c:107]:
```suggestion
    NULL /* encapsulate */,
```

File: crypto/evp_extra/p_x25519.c [hunk 3]
```diff
@@ -110,4 +110,6 @@ const EVP_PKEY_METHOD x25519_pkey_meth = {
     pkey_x25519_derive,
     NULL /* paramgen */,
     pkey_x25519_ctrl,
+    NULL /* encapuslate */,
```
Comment (torben-hansen) [inline crypto/evp_extra/p_x25519.c:113]:
```suggestion
    NULL /* encapsulate */,
```

File: crypto/evp_extra/p_x25519.c [hunk 4]
```diff
@@ -110,4 +110,6 @@ const EVP_PKEY_METHOD x25519_pkey_meth = {
...
+    NULL /* decapuslate */,
```
Comment (torben-hansen) [inline crypto/evp_extra/p_x25519.c:114]:
```suggestion
    NULL /* decapsulate */,
```

File: crypto/evp_extra/pq_kem_kat_tests_kyber512.cc [hunk 5]
```diff
@@ -13,10 +13,10 @@ static void RunTest(FileTest *t)
   std::string count;
```
Comment (torben-hansen) [inline crypto/evp_extra/pq_kem_kat_tests_kyber512.cc:13]: I don't think we should call this file `pq_`. I would rather we regard these "new" algorithms as any other algorithms without all the "pq"ing.

File: crypto/fipsmodule/evp/internal.h [hunk 6]
```diff
@@ -250,6 +239,17 @@ struct evp_pkey_method_st {
   int (*paramgen)(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey);
 
   int (*ctrl)(EVP_PKEY_CTX *ctx, int type, int p1, void *p2);
+
+  // Encapsulate and decapsulate are operations defined for a KEM
+  // (Key Encapsulation Mechanism). So far, they are used only for
```
Comment (torben-hansen) [inline crypto/fipsmodule/evp/internal.h:244]: IMO delete
> So far, they are used only for Kyber post-quantum KEM.

File: crypto/kyber/kem_kyber.h [hunk 7]
```diff
@@ -6,29 +6,21 @@
 #include <openssl/base.h>
 #include <openssl/evp.h>
 
-#define KYBER512_PRIVATE_KEY_BYTES 1632
 #define KYBER512_PUBLIC_KEY_BYTES 800
-#define KYBER512_KEM_CIPHERTEXT_BYTES 768
-#define KYBER512_KEM_SHARED_SECRET_BYTES 32
+#define KYBER512_SECRET_KEY_BYTES 1632
```
Comment (torben-hansen) [inline crypto/kyber/kem_kyber.h:10]: Reference where these values are taken from

File: crypto/kyber/p_kyber.c [hunk 8]
```diff
@@ -29,6 +29,53 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
...
 }
 
+static int pkey_kyber512_encapsulate(EVP_PKEY_CTX *ctx,
+                                     uint8_t *ciphertext,
+                                     size_t  *ciphertext_len,
+                                     uint8_t *shared_secret,
+                                     size_t  *shared_secret_len) {
+  *ciphertext_len    = KYBER512_CIPHERTEXT_BYTES;
+  *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
+  if (ciphertext == NULL) { // Caller is performing a size check
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber.c:39]: I guess the caller is really getting the parameter values, not performing a "check"?

File: crypto/kyber/p_kyber.c [hunk 9]
```diff
@@ -29,6 +29,53 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
   return 1;
 }
 
+static int pkey_kyber512_encapsulate(EVP_PKEY_CTX *ctx,
+                                     uint8_t *ciphertext,
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber.c:33]: Okay, the typical style is not to add newlines for every parameter. But it's not like I'm going to complain a lot about it. Just adding it as a note.

File: crypto/kyber/p_kyber.c [hunk 10]
```diff
@@ -29,6 +29,53 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
...
+  *ciphertext_len    = KYBER512_CIPHERTEXT_BYTES;
+  *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
+  if (ciphertext == NULL) { // Caller is performing a size check
+    return 1;
+  }
+
+  if (ctx != NULL &&
+      ctx->pkey != NULL &&
+      ctx->pkey->pkey.ptr != NULL &&
+      ctx->pkey->type == EVP_PKEY_KYBER512) {
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber.c:46]: I think we should try and branch on errors, not success. Unless it totally does not make sense.

File: crypto/kyber/p_kyber_test.cc [hunk 11]
```diff
@@ -141,21 +141,21 @@ TEST(Kyber512Test, KeySize) {
...
-  EXPECT_EQ(KYBER512_PUBLIC_KEY_BYTES + KYBER512_PRIVATE_KEY_BYTES, EVP_PKEY_size(kyber_pkey));
-  EXPECT_EQ(8*(KYBER512_PUBLIC_KEY_BYTES + KYBER512_PRIVATE_KEY_BYTES), EVP_PKEY_bits(kyber_pkey));
+  EXPECT_EQ(KYBER512_PUBLIC_KEY_BYTES + KYBER512_SECRET_KEY_BYTES, EVP_PKEY_size(kyber_pkey));
+  EXPECT_EQ(8*(KYBER512_PUBLIC_KEY_BYTES + KYBER512_SECRET_KEY_BYTES), EVP_PKEY_bits(kyber_pkey));
 
   EVP_PKEY_CTX_free(kyber_pkey_ctx);
 }
 
 TEST(Kyber512Test, KEMOperations) {
   // Basic functional test for KYBER512
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber_test.cc:151]: Just because now we are here :)
```suggestion
  // Basic functional test for KYBER512.
```

File: include/openssl/evp.h [hunk 12]
```diff
@@ -691,20 +691,23 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
 // containing the result. It returns one on success or zero on error.
 OPENSSL_EXPORT int EVP_PKEY_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY **out_pkey);
 
-OPENSSL_EXPORT int EVP_PKEY_encapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
+// EVP_PKEY_encapsulate is operation defined for a KEM (Key Encapsulation
```
Comment (torben-hansen) [inline include/openssl/evp.h:694]:
```suggestion
// EVP_PKEY_encapsulate is an operation defined for a KEM (Key Encapsulation
```

File: include/openssl/evp.h [hunk 13]
```diff
@@ -691,20 +691,23 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
...
-                                        unsigned char *gen_key,
-                                        size_t *gen_key_len);
-OPENSSL_EXPORT int EVP_PKEY_decapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
+                                        uint8_t *ciphertext,
+                                        size_t *ciphertext_len,
+                                        uint8_t *shared_secret_key,
+                                        size_t *shared_secret_len);
+
+// EVP_PKEY_decapsulate is operation defined for a KEM (Key Encapsulation
```
Comment (torben-hansen) [inline include/openssl/evp.h:703]:
```suggestion
// EVP_PKEY_decapsulate is an operation defined for a KEM (Key Encapsulation
```

File: include/openssl/evp.h [hunk 14]
```diff
@@ -691,20 +691,23 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
 // containing the result. It returns one on success or zero on error.
 OPENSSL_EXPORT int EVP_PKEY_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY **out_pkey);
 
-OPENSSL_EXPORT int EVP_PKEY_encapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
+// EVP_PKEY_encapsulate is operation defined for a KEM (Key Encapsulation
+// Mechanism). The function generates a shared secret and a ciphertext which
+// encapsulates the shared secret using the public key given within |ctx|.
 OPENSSL_EXPORT int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
```
Comment (torben-hansen) [inline include/openssl/evp.h:697]: Return values?
Where are the "ciphertext" and "shared secret" output?
What is `secret_key_len` exactly?

I think we might as well write it.

File: include/openssl/evp.h [hunk 15]
```diff
@@ -691,20 +691,23 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
...
-                                             const OSSL_PARAM params[]);
+                                        uint8_t *ciphertext,
+                                        size_t *ciphertext_len,
+                                        uint8_t *shared_secret_key,
+                                        size_t *shared_secret_len);
+
+// EVP_PKEY_decapsulate is operation defined for a KEM (Key Encapsulation
+// Mechanism). The function decapsulates the shared secret from the given
+// ciphertext using the private key given within |ctx|.
 OPENSSL_EXPORT int EVP_PKEY_decapsulate(EVP_PKEY_CTX *ctx,
```
Comment (torben-hansen) [inline include/openssl/evp.h:706]: Same as for `EVP_PKEY_encapsulate`. In addition, document what happens if `shared_secret = NULL`.

File: crypto/kyber/p_kyber.c [hunk 16]
```diff
@@ -29,6 +29,53 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
...
+    if (key->has_private) {
+      kyber512_decapsulate(shared_secret, ciphertext, key->priv);
+      return 1;
+    }
+  }
+
+  return 0;
+}
+
 const EVP_PKEY_METHOD kyber512_pkey_meth = {
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber.c:79]: So, are we going to have one implementation per parameter set? Can't this blow-up in our face at some point since there are likely going to be several parameter sets standardised for each "new" algorithm?

File: include/openssl/evp.h [hunk 17]
```diff
@@ -691,20 +691,23 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
...
+// Mechanism). The function generates a shared secret and a ciphertext which
+// encapsulates the shared secret using the public key given within |ctx|.
 OPENSSL_EXPORT int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
-                                        unsigned char *wrapped_key,
-                                        size_t *wrapped_key_len,
-                                        unsigned char *gen_key,
-                                        size_t *gen_key_len);
-OPENSSL_EXPORT int EVP_PKEY_decapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
+                                        uint8_t *ciphertext,
```
Comment (andrewhop) [inline include/openssl/evp.h:698]: Why the change to diverge from OpenSSL? https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_decapsulate.html uint8_t line up with what we use elsewhere, but we should have clear API guidelines for when we diverge or not. How loudly do compilers complain about customer code casting an unsigned char to uint8_t?

File: include/openssl/evp.h [hunk 18]
```diff
@@ -691,20 +691,23 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
...
+                                        size_t *ciphertext_len,
+                                        uint8_t *shared_secret_key,
```
Comment (andrewhop) [inline include/openssl/evp.h:700]: Did you consider naming these `ciphertext_len_out` to make it a bit more clear? Or is already clear that encapsulate will produce a ciphertext and shared_secret_key?

File: include/openssl/evp.h [hunk 19]
```diff
@@ -691,20 +691,23 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
 // containing the result. It returns one on success or zero on error.
 OPENSSL_EXPORT int EVP_PKEY_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY **out_pkey);
 
-OPENSSL_EXPORT int EVP_PKEY_encapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
+// EVP_PKEY_encapsulate is an operation defined for a KEM (Key Encapsulation
+// Mechanism). The function generates a shared secret and a ciphertext which
+// encapsulates the shared secret using the public key given within |ctx|.
```
Comment (andrewhop) [inline include/openssl/evp.h:696]: TBD word smiting, can this be a bit more clear what is written out vs passed into this function, something like:
```suggestion
// EVP_PKEY_encapsulate is an operation defined for a KEM (Key Encapsulation
// Mechanism). The function generates a shared secret, writes it to |shared_secret|, and
// encapsulates with the public key in |ctx| and written to |ciphertext|.
```

Also can you update it to specify what is allocated vs expected to already be allocated

**Comment by dkostic** [comment]
closing until an internal review is done.
